tdi_python/tdiTable.py

```diff
@@ -0,0 +1,2617 @@
+        return sts
+
+    """
+    A helper function for adding or modifying table entries, as these
+    two operations are essentially the same - but an add must have been
+    called on a key before said key can be modified.
+    """
+    def _call_add_mod_inc(self, key_content, data_content, action, flag_type, c_func):
+        key_handle = self._make_call_keys(key_content)
+        data_handle = self._make_call_data(data_content, action)
+        if key_handle == -1 or data_handle == -1:
+            # If one was -1 but the other was created, then it must be freed
+            if key_handle != -1:
+                self._cintf.get_driver().tdi_table_key_deallocate(key_handle)
+            if data_handle != -1:
+                self._cintf.get_driver().tdi_table_data_deallocate(data_handle)
+            return -1
+        sts = c_func(self._handle, self._cintf.get_session(), self._cintf.get_dev_tgt(), key_handle, data_handle, flag_type)
+        self._cintf.get_driver().tdi_table_key_deallocate(key_handle)
+        self._cintf.get_driver().tdi_table_data_deallocate(data_handle)
+
+        return sts
+
+    """
+    These methods essentially are the highest level of abstraction on our
+    BF Runtime C API. All data passed to them must be eligible for use with
+    Python's ctypes foriegn function interface.
+    """
+    def add_entry(self, key_content, data_content, action=None):
+        sts = self._call_add_mod(key_content, data_content, action, self._cintf.tdi_table_entry_add)
+        if not sts == 0:
+            raise TdiTableError("Error: table_entry_add failed on table {}. [{}]".format(self.name, self._cintf.err_str(sts)), self, sts)
+
+    def mod_entry(self, key_content, data_content, action=None, ttl_reset=True):
+        flags = 0
+        if ttl_reset == False:
+            # SKIP_TTL_RESET is second bit, currently only reset supported.
+            flags |= 1 << 2
+        sts = self._call_add_mod(key_content, data_content, action, self._cintf.tdi_table_entry_mod, flags)
+        if not sts == 0:
+            raise TdiTableError("Error: table_entry_mod failed on table {}. [{}]".format(self.name, self._cintf.err_str(sts)), self, sts)
+
+    '''
+    def mod_inc_entry(self, key_content, data_content, flag_type=0, action=None):
+        sts = self._call_add_mod_inc(key_content, data_content, action, flag_type, self._cintf.tdi_table_entry_mod_inc)
+        if not sts == 0:
+            raise TdiTableError("Error: table_entry_mod failed on table {}. [{}]".format(self.name, self._cintf.err_str(sts)), self, sts)
+    '''
+
+    def set_default_entry(self, data_content, action=None):
+        table_type = self.table_type_map(self.get_type())
+        if table_type == -1:
+            return -1
+        data_handle = self._make_call_data(data_content, action)
+        if data_handle == -1:
+            return -1
+        sts = self._cintf.tdi_table_default_entry_set(self._handle,
```

Comment:
we should be calling the API version which takes in flags and not these older versions